proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+```json
+{
+    "json-rpc":"2.0",
+    "result":{ "res":"value" },
+    "id":1
+}
+```
+#### JSON Response (error) - contains error code
+```json
+{
+    "jsonrpc": "2.0",
+     "error": {"code": -32601, "message": "Method not found"},
+     "id": 1
+}
+```
+
+### Protocol communication with example JSON-RPC messages
+A simple compilation with no include handler callback provided will be
+implemented as a single `Request` followed by a single `Response`.
+
+```
+REQUEST  - "compile"
+RESPONSE - "result", compilation succeeded/failed.
+```
+
+```
+Syntax:
+--> data sent to server
+<-- data coming from server
+```
+
+### Example: compile, no include handler, success
+```
+--> {"jsonrpc": "2.0", "method": "compile",
+     "params": {"arg1": 23, "arg2": 42}, "id": 1}
+
+<-- {"jsonrpc": "2.0", "result": {"stdout": "somepath", "stderr": "somepath"},
+     "id": 1}
+```
+
+### Example: compile, no include handler, failure
+```
+--> {"jsonrpc": "2.0", "method": "compile",
+     "params": {"arg1": 23, "arg2": 42}, "id": 1}
+
+<-- {"jsonrpc": "2.0", "error": {"code": -2, "message": "compile failed"},
+     "data": {"stdout": "somepath", "stderr": "somepath"}, "id": 1}
+```
+
+Include handlers add some additional complexity but still keep to the
+`Request`/`Response` pair design.  If an include handler callback is provided
+to the c api, a special proxy implementation of that callback is created and 
+connected to the IPC messaging system and lives in the clang.exe process.
+This proxy callback will be used during compilation when an include is
+requested.  Calling the proxy callback will result in IPC communication with
+the Api Dispatcher in the form of a `Response` indicating an include is needed.
+
+A compilation is started using a single "compile" `Request` and additional state
+is provided that indicates include handler support is available.
+
+Compilation proceeds and blocks the originating `Request` until the first need
+an include handler is encountered. The proxy callback is called as the include
+handler and the implementation sends a `Response` back to the Api Dispatcher
+indicating that include handler information is needed. This will close the
+originating "compile" `Request`/`Response` pair.
+
+The Api Dispatcher code (in the caller's process) sees this `Response` and
+immediately calls the c api provided include handler callback. The results are
+collected and new "continue compilation" `Request` is dispatched.
+
+The worker process (currently waiting for another `Request` to continue)
+receives the `Request` with include handler data as parameters and proceeds
+with compilation.
+
+This `Request`/`Response` pattern continues until the compilation is completed.
+Compilation is always concluded with a final `Response` closing the
+`Request`/`Response` pair.
+
+```
+REQUEST  - "compile" (include handler support is enabled)
+
+( compiler needs an include, calls IPC wrapped included handler callback )
+
+RESPONSE - "needinclude"
```

Comment:
This "needinclude" response feels more like clang wanting to call back into the calling process.  Any reason not to model this in that way, having a more consistent REQUEST/RESPONSE matching to a function and a return value?

eg:

```
---> REQUEST  id:1 compile...
<--- REQUEST  id:2 need include...
---> RESPONSE id:2 include result
<--- REQUEST  id:3 need include...
---> RESPONSE id:3 include result
<--- RESPONSE id:1 compilation result
````

Reply:
I like this idea.  I misunderstood the json-rpc protocol on my first read through.  I thought the server was not allowed to send a request/response pair to the client, but I was wrong. That is totally supported.